File: 01-arrays-and-strings/03_trapping_rain_water/TrappingRainWater.java
/**
 * Problem: Trapping Rain Water
 * LeetCode: #42 (Hard)
 * 
 * Approach: Two Pointers
 * Time Complexity: O(N)
 * Space Complexity: O(1)
 */

public class TrappingRainWater {
    public static int trap(int[] height) {
        int left = 0, right = height.length - 1;
        int leftMax = 0, rightMax = 0;
        int totalWater = 0;

        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] >= leftMax) {
                    leftMax = height[left];
                } else {
                    totalWater += leftMax - height[left];
                }
                left++;
            } else {
                if (height[right] >= rightMax) {
                    rightMax = height[right];
                } else {
                    totalWater += rightMax - height[right];
                }
                right--;
            }
        }
        return totalWater;
    }

    public static void main(String[] args) {
        System.out.println("[Java] Trapping Rain Water Test");
        int[] height = { 0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1 };
        System.out.println("Trapped Water: " + trap(height) + " (Expected: 6)");
    }
}

File: 01-arrays-and-strings/03_trapping_rain_water/trapping_rain_water.py
"""
Problem: Trapping Rain Water
LeetCode: #42 (Hard)

Approach: Two Pointers
Time Complexity: O(N)
Space Complexity: O(1)
"""

from typing import List

def trap(height: List[int]) -> int:
    left, right = 0, len(height) - 1
    left_max, right_max = 0, 0
    total_water = 0

    while left < right:
        if height[left] < height[right]:
            if height[left] >= left_max:
                left_max = height[left]
            else:
                total_water += left_max - height[left]
            left += 1
        else:
            if height[right] >= right_max:
                right_max = height[right]
            else:
                total_water += right_max - height[right]
            right -= 1

    return total_water

if __name__ == "__main__":
    print("[Python] Trapping Rain Water Test")
    tests = [
        ([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1], 6),
        ([4, 2, 0, 3, 2, 5], 9),
    ]
    for height, expected in tests:
        res = trap(height)
        print(f"height: {height} -> Trapped: {res} | Expected: {expected}")
        assert res == expected
    print("All Trapping Rain Water tests passed!")
